Fixed #33422 -- Improved docs about isolating apps. (#15735)
django/django: docs/topics/testing/advanced.txt

    configuration that need to be reversed. It's the return value of the
    :meth:`setup_databases` method.

.. function:: isolate_apps(*app_labels, attr_name=None, kwarg_name=None)

    Register the models defined within a wrapped context into their own

>     isolated app registry. This functionality is useful when creating a

Reviewer: link to app registry class and mention it by name, `Apps`, as per the existing contributing docs description

Author: 👍

    Django model for a test case. An isolated registry avoids mixing
    concerns as well as edge cases involving name collision.

    The app labels which the isolated registry should contain must be
    passed as individual arguments. And the invocation can be either a

>     decorator or a context manager. For example::

Reviewer: not great to start a sentence with "and". perhaps:

"You can use ``isolate_apps`` as a decorator or a context manager."

Author: 👍

        class MyModelTests(SimpleTestCase):

            @isolate_apps("app_label")
            def test_model_definition(self):
                class TestModel(models.Model):
                    pass
                ...

    Or::

        with isolate_apps("app_label"):
            class TestModel(models.Model):
                pass
            ...

>     See also, :ref:`isolating-model-registration`.

Reviewer: I'm not sure about the main django docs linking to the contributing docs. It could be confusing for users to find themselves with loads of other information about "how you should be doing things" that's more specific to django's code base. Better to move any useful information from that section here, and leave the contributing docs with just a pointer to `isolate_apps`.

Author: 👍

    configuration that need to be reversed. It's the return value of the
    :meth:`setup_databases` method.

> .. function:: isolate_apps(*app_labels, attr_name=None, kwarg_name=None)

Reviewer 2: I don't think it's the right place to document `@solate_apps`. These functions are described as:
```
To assist in the creation of your own test runner, Django provides a number of
utility methods in the ``django.test.utils`` module.
```
I would move it to the new paragraph e.g. "Isolating apps" in the ["Test cases features"](https://docs.djangoproject.com/en/dev/topics/testing/tools/#test-cases-features).

    Registers the models defined within a wrapped context into their own

>     isolated :attr:`~django.apps.apps` registry. This functionality is
>     useful when creating a model for a test case. An isolated
>     registry avoids mixing concerns as well as edge cases involving name
>     collision.

Reviewer: Re-reading this, "mixing concerns" is kinda vague, I tried to spell it out:

```suggestion
    isolated :attr:`~django.apps.apps` registry. This functionality is
    useful when creating model classes for tests, as the classes will
    be cleanly deleted afterwards, and there is no risk of name
    collisions.
```

    The app labels which the isolated registry should contain must be
    passed as individual arguments. You can use ``isolate_apps()`` as

>     a decorator or a context manager. For example::

Reviewer 2: Wrap at 79 chars.

        class MyModelTests(SimpleTestCase):

            @isolate_apps("app_label")
            def test_model_definition(self):
                class TestModel(models.Model):
                    pass
                ...

    Or::

        with isolate_apps("app_label"):
            class TestModel(models.Model):
                pass
            ...

    The decorator form can also be applied to classes.

    Two optional keyword arguments can be specified:

>         * ``attr_name``: attribute assigned the isolated registry if
>           used as a class decorator.
>         * ``kwarg_name``: keyword argument passing the isolated registry
>           if used as a function decorator.

Reviewer 2: ```suggestion
    * ``attr_name``: attribute assigned the isolated registry if used as a
      class decorator.
    * ``kwarg_name``: keyword argument passing the isolated registry if used as
      a function decorator.
```

    retrieved as an attribute when used as a class decorator by using the
    ``attr_name`` parameter::

        from django.db import models
        from django.test import SimpleTestCase
        from django.test.utils import isolate_apps

        @isolate_apps('app_label', attr_name='apps')
        class TestModelDefinition(SimpleTestCase):
            def test_model_definition(self):
                class TestModel(models.Model):
                    pass
                self.assertIs(self.apps.get_model('app_label', 'TestModel'), TestModel)

    Or alternatively as an argument on the test method when used as a
    method decorator by using the ``kwarg_name`` parameter::

        from django.db import models
        from django.test import SimpleTestCase
        from django.test.utils import isolate_apps

        class TestModelDefinition(SimpleTestCase):
            @isolate_apps('app_label', kwarg_name='apps')
            def test_model_definition(self, apps):
                class TestModel(models.Model):
                    pass
                self.assertIs(apps.get_model('app_label', 'TestModel'), TestModel)

> 

Reviewer 2: Chop blank line
```suggestion
```
